From 2a60664e7bae41f878892ae8156571bef3edd99c Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Fri, 4 Sep 2015 13:01:07 +0530
Subject: [PATCH] USBC: Cancel SrcCap/DI workers without holding mutex.

As SrcCap and DI resending workers using muex lock, cancel_delayed_work_sync()
with mutex locked will cause deadlock. Hence cancel worker is called
without holding mutex to allow the worker complete it's task.

Change-Id: I1137198de9bfdb55b2ceec7bf2f1edf9e63bbfda
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-17935
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
---
 drivers/usb/typec/pd/display_pe.c  |    4 ++--
 drivers/usb/typec/pd/src_port_pe.c |    7 +------
 2 files changed, 3 insertions(+), 8 deletions(-)

Index: b/drivers/usb/typec/pd/display_pe.c
===================================================================
--- a/drivers/usb/typec/pd/display_pe.c	2016-03-08 15:35:55.267630713 -0800
+++ b/drivers/usb/typec/pd/display_pe.c	2016-03-08 15:36:21.355953237 -0800
@@ -101,10 +101,10 @@
 
 static void disp_pe_reset_policy_engine(struct disp_port_pe *disp_pe)
 {
+	cancel_delayed_work_sync(&disp_pe->start_comm);
 	mutex_lock(&disp_pe->pe_lock);
 	disp_pe->p.state = POLICY_STATE_OFFLINE;
 	disp_pe->p.status = POLICY_STATUS_UNKNOWN;
-	cancel_delayed_work(&disp_pe->start_comm);
 	disp_pe->cmd_retry = 0;
 	disp_pe->dp_mode = TYPEC_DP_TYPE_NONE;
 	if (disp_pe->hpd_state) {
@@ -127,7 +127,7 @@
 static void disp_pe_handle_dp_fail(struct disp_port_pe *disp_pe)
 {
 	log_info("DP failed\n");
-	cancel_delayed_work(&disp_pe->start_comm);
+	cancel_delayed_work_sync(&disp_pe->start_comm);
 	mutex_lock(&disp_pe->pe_lock);
 	disp_pe->state = DISP_PE_STATE_ALT_MODE_FAIL;
 	disp_pe->p.status = POLICY_STATUS_FAIL;
Index: b/drivers/usb/typec/pd/src_port_pe.c
===================================================================
--- a/drivers/usb/typec/pd/src_port_pe.c	2016-03-08 15:36:21.355953237 -0800
+++ b/drivers/usb/typec/pd/src_port_pe.c	2016-03-08 15:38:24.857477038 -0800
@@ -968,19 +968,14 @@
 					struct src_port_pe, p);
 
 	log_info("IN");
+	cancel_delayed_work_sync(&src_pe->start_comm);
 	mutex_lock(&src_pe->pe_lock);
 	p->state = POLICY_STATE_OFFLINE;
 	p->status = POLICY_STATUS_UNKNOWN;
 	src_pe_reset_policy_engine(src_pe);
-	cancel_delayed_work_sync(&src_pe->start_comm);
 	policy_set_pd_state(p, false);
 	src_pe->cmd_retry = 0;
-	src_pe->got_snk_caps = 0;
-	src_pe->is_pd_configured = 0;
 	src_pe->vbus_retry_cnt = 0;
-	src_pe->pp_is_dual_drole = 0;
-	src_pe->pp_is_dual_prole = 0;
-	src_pe->pp_is_ext_pwrd = 0;
 	mutex_unlock(&src_pe->pe_lock);
 	return 0;
 }
